File: src/main/java/no/novari/configuration/ConfigurationRepository.java
package no.novari.configuration;

import no.novari.configuration.model.configuration.entities.Configuration;
import org.springframework.data.jpa.repository.JpaRepository;

import java.util.Optional;

public interface ConfigurationRepository extends JpaRepository<Configuration, Long> {

    default Configuration saveWithVersion(Configuration configuration) {
        if (configuration.isCompleted()) {
            int nextVersion = getNextVersionForIntegrationId(configuration.getIntegrationId());
            configuration.setVersion(nextVersion);
        }
        return save(configuration);
    }

    default int getNextVersionForIntegrationId(Long integrationId) {
        return findFirstByIntegrationIdAndVersionNotNullOrderByVersionDesc(integrationId)
                .map(Configuration::getVersion)
                .map(version -> version + 1)
                .orElse(1);
    }

    Optional<Configuration> findFirstByIntegrationIdAndVersionNotNullOrderByVersionDesc(Long integrationId);

}

File: src/main/kotlin/no/novari/flyt/configuration/ConfigurationRepository.kt
package no.novari.flyt.configuration

import no.novari.flyt.configuration.model.configuration.entities.Configuration
import org.springframework.data.jpa.repository.JpaRepository

interface ConfigurationRepository : JpaRepository<Configuration, Long> {
    fun saveWithVersion(configuration: Configuration): Configuration {
        if (configuration.isCompleted) {
            val nextVersion = getNextVersionForIntegrationId(requireNotNull(configuration.integrationId))
            configuration.version = nextVersion
        }
        return save(configuration)
    }

    fun getNextVersionForIntegrationId(integrationId: Long): Int =
        findFirstByIntegrationIdAndVersionNotNullOrderByVersionDesc(integrationId)
            ?.version
            ?.plus(1)
            ?: 1

    fun findFirstByIntegrationIdAndVersionNotNullOrderByVersionDesc(integrationId: Long): Configuration?
}
